### ISSUE TITLE
Add public method to Page control that ensures all TwoWay bindings are updated. (Form Submit)

### ISSUE THREAD

**Issue by omnilogix** [microsoft/microsoft-ui-xaml#5473]
# Proposal: Add pubic method to Page control that ensures all TwoWay bindings are updated. (Form Submit) 

## Summary

Add a method or equivalent feature that ensures that all TwoWay bindings on a Page or UserControl etc. are updated.  This would be called during Form submission to ensure the ViewModel is in sync with the user input at time of validation and data save. 

## Rationale

* Current UI design allows for a button click or keyboard accelerator that does not change the current focus.  Without focus change, TextBox default UpdateSourceTrigger of FocusChanged will not fire, and data binding remains incomplete for the currently focused input, allowing for un-validated as well as un-saved user data.
* In previous frameworks (WPF, etc), the binding could (with significant effort) be programmatically discovered and updates could be forced when necessary.  With x:Bind, this is no longer possible.
* This not-so-edge case has been overlooked in all previous frameworks
* One might argue that UpdateSourceTrigger could be set to PropertyChanged.  While true, this is far from optimal, especially when performing validation which would usually prefer to be on FocusChanged event, not after every keystroke.
* Now is the time to include such functionality.  Before GA release
* Getting the currently focused control and trying to move the focus is unreliable at best.  There should be a way to explicity update the bindings, or implicitly through a feature that would trigger a focus change event on the current control

## Scope

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers to accomplish Proper data binding | Must |
| This proposal will allow end users to be sure their input data is validated and saved | Should |

**Comment by mrlacey** [comment]
Why only do this on a `Page` control? Wouldn't it also make sense to have this on a `Window` or any custom control too?

If this is specifically related to Forms, should this be connected to #82 and as it has connections to validation, a consideration for #179 may also be appropriate.

The idea of "accomplish Proper data binding" is vague and potentially judgmental. Is the goal of avoiding workarounds and seemingly unnecessary extra code that is also hard to test also appropriate.

Would it be sufficient to have a button press also move focus? (I suspect not, but this needs exploring and documenting.)

Even doing the proposed seems like only a potential partial solution. As a keyboard accelerator or button click can be used to invoke a command that might start a form submission process, there's no guarantee that the form is also being updated to reflect that a background operation is in progress. A clear way of saying "the form should go into submission mode" is needed. This would force all binding updates, so validation can be done and then the UI appropriately updated to indicate that something is happening while any background operation is done. (hence the links to the issues mentioned above)

**Comment by omnilogix** [comment]
@mrlacey In the broader scope, yes, this feature should be available for any hosting control, but probably most useful as a feature of the binding system itself.  If we can call "UpdateBindings(..)", then we don't have to be concerned about the current state of the focus, managing focus, or even which kind of control currently holds the focus.  My particular concern is for data validation in LOB type apps.  I'm sure there are many other scenarios that would benefit.

**Comment by StephenLPeters** [comment]
@marb2000 @ryandemopoulos @MikeHillberg Seems like a good idea.

**Comment by MikeHillberg** [comment]
Love to have some form of this too (no pun intended).

Two existing precedences to think about
* WPF has [BindingGroup](https://docs.microsoft.com/dotnet/api/System.Windows.Data.BindingGroup) for this purpose, which is a single object that represents all of the {Bindings} in scope. The `CommitEdit()` method on that flushes all the TwoWay bindings back to source.
* x:Bind in WinUI provides an `Update()` method which unfortunately goes in the opposite direction from the case here. That's useful too but not the scenario proposed here. There's not a lot of documentation but there is some [here](https://docs.microsoft.com/en-us/windows/uwp/xaml-platform/x-bind-markup-extension#remarks). So on your page/usercontrol/etc you can call `this.Bindings.Update();` to get all of the binds to update from source to the view.

**Comment by chrisglein** [comment]
Agreed that from a scenario perspective this feels like part of the need for a Forms control.

> If this is specifically related to Forms, should this be connected to #82 and as it has connections to validation, a consideration for #179 may also be appropriate.

So +1 to that.

It does sound like there might be some binding features from WPF that could be relevant, independent from the Forms scenario:

> WPF has BindingGroup for this purpose, which is a single object that represents all of the {Bindings} in scope. The CommitEdit() method on that flushes all the TwoWay bindings back to source.

So we can keep this issue open to track that class of feature. @omnilogix it would be helpful if you could create a small illustrative repro of exactly what you'd hope to get here, that could be used as the benchmark of whether the binding feature meets the needs or not.

But to set expectations on this:

> Now is the time to include such functionality. Before GA release

WinUI's first job is to replicate the experience of UWP XAML, lifted outside of the OS. That's what's generally expected for initial release, and there will still be gaps between those. Additional gaps between UWP and WPF are tracked (see [these issues](https://github.com/microsoft/microsoft-ui-xaml/labels/wpf-vs-winui-mismatch)) to potentially be tackled over time.

**Comment by omnilogix** [comment]
@chrisglein I'll work on getting a repo together, but I am a little short on time right now.  In summary though, The heart of the issue is that the currently focused control does not update it's binding in all form-submit scenarios.  The WPF BindingGroup creates the necessity to add code for all potential TwoWay bound controls which is in my opinion adds too much complexity.  All that is really necessary is a way to Trigger a binding update for the control that currently holds the focus.  Where exactly is the best place to plumb that in, I am not sure.  I do know that a "FlushTwoWayBindings()" method would solve all of the issues that I deal with as far as forms submission go.  I really don't feel that this is a "Forms" issue so much as a binding issue.  Just my 2 cents.

**Comment by MikeHillberg** [comment]
It's a binding issue that's common to the forms scenario, but you're right it's not limited to forms.

Can you do this today using [FrameworkElement.GetBindingExpression](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.FrameworkElement.GetBindingExpression) to get the binding from the TextBox.Text property, and then call [BindingExpression.UpdateSource](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Data.BindingExpression.UpdateSource) on it to flush the value?

**Comment by omnilogix** [comment]
> It's a binding issue that's common to the forms scenario, but you're right it's not limited to forms.
> 
> Can you do this today using [FrameworkElement.GetBindingExpression](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.FrameworkElement.GetBindingExpression) to get the binding from the TextBox.Text property, and then call [BindingExpression.UpdateSource](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Data.BindingExpression.UpdateSource) on it to flush the value?

@MikeHillberg Not with the x:Bind, which I am using primarily

**Comment by MikeHillberg** [comment]
Got it. That's where I'd like to go too. For x:Bind we have `this.Bindings.Update()` to update from source to target, what we need is a `this.Bindings.UpdateSource()` to go in the other direction.

**Comment by omnilogix** [comment]
@MikeHillberg Is the this.Bindings.UpdateSource() available for WinUI?  I didn't find it with the Page Control.  The docs looked like it was for UWP.  Thanks

**Comment by MikeHillberg** [comment]
You're right, it's not. I was suggesting it as a solution.

**Comment by oliverw** [comment]
> You're right, it's not. I was suggesting it as a solution.

So what are you supposed to do call in WinUI to force refresh the bindings?

**Comment by SoggyBottomBoy** [comment]
I have a similar problem which I believe is related to this. Currently I have a list view, which is a list of customer viewmodels, the selected item is bound to a property `SelectedCustomer` in my ViewModel for the page. I also have a form on the page which is bound to the SelectedCustomer. 

The issue I am having it that if the user types something in a textbox/number box doesn't click/tab away (i.e. the control still has focus) then clicks on another item in the list view the binding to source does not get triggered and the text box carries the newly typed value over to the new selected view model and updates the newly selected view model.

I thought my solution might be to manually trigger a Bindings.Update() when an item in the list view is clicked to force an update. But as described above I need view to source, not source to view.

Is there a way to trigger a bindings view to source via another mechanism (I have tried forcing textboxes/number box to lose focus, but this doesn't seem to work).

**Comment by omnilogix** [comment]
Has there been any movement on this issue?
